cmd/juju/interact/pollster.go

```diff
@@ -0,0 +1,515 @@
+// from a list of options.
+type List struct {
+	Singular string
+	Plural   string
+	Options  []string
+	Default  string
+}
+
+// MultiList contains the information necessary to ask the user to select form a list
+// of options.
+type MultiList struct {
+	Singular string
+	Plural   string
+	Options  []string
+	Default  []string
+}
+
+var listTmpl = template.Must(template.New("").Funcs(map[string]interface{}{"title": strings.Title}).Parse(`
+{{title .Plural}}
+{{range .Options}}  {{.}}
+{{end}}
+`[1:]))
+
+var selectTmpl = template.Must(template.New("").Parse(`
+Select {{.Singular}}{{if .Default}} [{{.Default}}]{{end}}: `[1:]))
+
+// Select queries the user to select from the give list of options.
+func (p *Pollster) Select(l List) (string, error) {
+	return p.SelectVerify(l, VerifyOptions(l.Singular, l.Options, l.Default != ""))
+}
+
+// SelectVerify queries the user to select from the give list of options, using
+// the custom verifier listed.
+func (p *Pollster) SelectVerify(l List, verify VerifyFunc) (string, error) {
+	if err := listTmpl.Execute(p.out, l); err != nil {
+		return "", err
+	}
+
+	question, err := sprint(selectTmpl, l)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	val, err := QueryVerify(question, p.scanner, p.out, p.errOut, verify)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	if val == "" {
+		return l.Default, nil
+	}
+	return val, nil
+}
+
+var multiSelectTmpl = template.Must(template.New("").Funcs(
+	map[string]interface{}{"join": strings.Join}).Parse(`
+Select one or more {{.Plural}} separated by commas{{if .Default}} [{{join .Default ", "}}]{{end}}: `[1:]))
+
+// MultiSelect queries the user to select one more answers from the give list of
+// options by entering values delimited by commas (and thus options must not
+// contain commas).
+func (p *Pollster) MultiSelect(l MultiList) ([]string, error) {
+	if err := listTmpl.Execute(p.out, l); err != nil {
+		return nil, err
+	}
+
+	question, err := sprint(multiSelectTmpl, l)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	a, err := QueryVerify(question, p.scanner, p.out, p.errOut, multiVerify(l.Singular, l.Plural, l.Options, l.Default != nil))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if a == "" {
+		return l.Default, nil
+	}
+	return multiSplit(a), nil
+}
+
+// Enter requests that the user enter a value.  Any value except an empty string
+// is accepted.
+func (p *Pollster) Enter(valueName string) (string, error) {
+	return p.EnterVerify(valueName, nil)
+}
+
+// VerifyFunc is a type that determines whether a value entered by the user is
+// acceptable or not.  If it returns an error, the calling func will return an
+// error, and the other return values are ignored.  If ok is true, the value is
+// acceptable, and that value will be returned by the calling function.  If ok
+// is false, the user will be asked to enter a new value for query.  If ok is
+// false. if errmsg is not empty, it will be printed out as an error to te the
```

Comment:
So.... error is an error - something went badly wrong, we should bail out.
ok is whether or not the string passes verification.
the message is what is shown if the message fails verification. 

The reason we can't just have a string and check for empty is that there are times when we want to fail verification but not print anything out.  For now, that basically only happens if the input is an empty string and there's no default.  Instead of printing out `"" is not a valid foo`, we just redisplay the query.